crates/bevy_transform/src/systems.rs

```diff
@@ -36,150 +45,197 @@ pub fn sync_simple_transforms(
 
-            for (child, actual_parent) in parent_query.iter_many(children) {
-                assert_eq!(
-                    actual_parent.get(), entity,
-                    "Malformed hierarchy. This probably means that your hierarchy has been improperly maintained, or contains a cycle"
+            // SAFETY: Visiting the hierarchy as a tree without cycles
+            #[expect(unsafe_code, reason = "Mutating disjoint entities in parallel")]
+            unsafe {
+                let mut outbox = Vec::new();
+                propagate_to_child_unchecked(
+                    (parent, parent_transform, children),
+                    &nodes,
+                    &mut outbox,
                 );
-                // SAFETY:
-                // - `child` must have consistent parentage, or the above assertion would panic.
-                // Since `child` is parented to a root entity, the entire hierarchy leading to it is consistent.
-                // - We may operate as if all descendants are consistent, since `propagate_recursive` will panic before 
-                //   continuing to propagate if it encounters an entity with inconsistent parentage.
-                // - Since each root entity is unique and the hierarchy is consistent and forest-like,
-                //   other root entities' `propagate_recursive` calls will not conflict with this one.
-                // - Since this is the only place where `transform_query` gets used, there will be no conflicting fetches elsewhere.
-                #[expect(unsafe_code, reason = "`propagate_recursive()` is unsafe due to its use of `Query::get_unchecked()`.")]
-                unsafe {
-                    propagate_recursive(
-                        &global_transform,
-                        &transform_query,
-                        &parent_query,
-                        child,
-                        changed || actual_parent.is_changed(),
-                    );
-                }
+                queue.sender.send(outbox).ok();
             }
-        },
-    );
+        });
+
+    // Spawn workers on the task pool to recursively propagate the hierarchy in parallel.
+    let task_pool = ComputeTaskPool::get_or_init(TaskPool::default);
+    task_pool.scope(|s| {
+        (0..task_pool.thread_num())
+            .for_each(|_| s.spawn(async { propagation_worker(&queue, &nodes) }));
+    });
 }
 
-/// Recursively propagates the transforms for `entity` and all of its descendants.
+/// A parallel worker that will consume processed parent entities from the queue, and push children
+/// to the queue once it has propagated their [`GlobalTransform`].
+fn propagation_worker(
+    queue: &WorkQueue,
+    nodes: &Query<(
+        Entity,
+        Ref<'_, Transform>,
+        &mut GlobalTransform,
+        &Children,
+        &ChildOf,
+    )>,
+) {
+    let mut outbox = queue.local_queue.borrow_local_mut();
+    loop {
+        // Try to acquire a lock on the work queue in a tight loop.
```

Comment:
I'm not too familiar with mutexes; would something like this https://marabos.nl/atomics/basics.html#condvar be benefitial to avoid a spinloop?

Reply:
Maybe. The workers should ever be waiting other than the very beginning, so there is likely no real benefit.

Reply:
To justify this claim, the gap between tasks (when spinning happens) is only300-500ns long
![image](https://github.com/user-attachments/assets/b38c82a8-9f36-4030-af54-b4d642dd6a12)

Reply:
I think if we intend to spin loop we can provide a hint to the compiler to make some optimizations: https://doc.rust-lang.org/std/hint/fn.spin_loop.html